### DISCUSSION TITLE
Mouse  cursor

### DISCUSSION THREAD

**Discussion by defdefred** [VendelinSlezak/BleskOS#39] [Ideas]
I wonder if saving the background of the cursor is indeed useful?
How about drawing directly the mouse to the framebuffer and using the double buffer as background saving?
You never need to draw the mouse to the double buffer...

**Comment by VendelinSlezak** [comment]
Thank you. I did not realize this, but yes, it is true that mouse cursor is not needed in double graphic buffer. I optimized your idea even more, so now is cursor drawed from methods for redrawing screen. So now BleskOS do not need to call mouse_cursor_save_background, mouse_cursor_restore_background and draw_mouse_cursor what saves space. Actual code could be optimized even more, but I will do it when I will have time.

**Comment by defdefred** [comment]
Are you sure that it is faster like this?
The "redraw_part_of_screen" function is way more complicated with branching slowing the cpu/cache usage.
Now the function is slower even when no mouse is involve in the part of the screen redrawn...

Why do you draw the mouse in the double buffer?
Just draw it to the frame buffer, after the redrawn, without even saving the background as it is already kept in the double buffer.

Trying with really old hardware could help testing optimization... or maybe displaying performance statistic on screen is possible? Could be a new program of bleskos to validate optimization for mouse drawing, character drawing, full screen redraw, etc :-).

I wonder if drawing the mouse to the frame-buffer after anytime the frame-buffer is redrawn shouldn't be faster than testing if we need to redraw the mouse or not... and the move of the mouse could be just one redraw_part_of_screen to the old position.

I wonder too if limiting the screen area accessible by the mouse to graphic_screen_y-MOUSE_CURSOR_HEIGTH and graphic_screen_x-MOUSE_CURSOR_WIDTH couldn't be a easy optimization for the drawing.
The bottom and right border of the screen could be dedicated for menu, button, scrollbar and activated by the mouse touching the limit of the mouse area and never outside the physical screen.

**Comment by VendelinSlezak** [comment]
I greatly optimized everything in Update 2, I am sure now it is faster then before, number of IF is reduced to minimum and also number of computation of variabiles. I draw mouse in double buffer, because otherwise there could be blinking between redrawing double buffer and then drawing mouse cursor directly to screen. So instead of IF statement if we are already on part where is mouse cursor, that would be executed in every pixel, actual code if we are redrawing part, where is mouse cursor, just draw mouse cursor to double buffer, redraw screen and then restore that part of double buffer.

**Comment by defdefred** [comment]
Interesting, could be worth testing if the mouse is blinking...
And maybe, not redrawing the mouse on full redraw could be a solution... the mouse just disapeer and came back on first mouse move...

I will look at your new code this evening :-)

**Comment by defdefred** [comment]
Like the function pointer usage :-)
